executor/sort.go

```diff
@@ -131,6 +146,153 @@ func (e *SortExec) Next() (Row, error) {
 	return row, nil
 }
 
+// NextChunk implements the Executor NextChunk interface.
+func (e *SortExec) NextChunk(chk *chunk.Chunk) error {
+	chk.Reset()
+	if !e.fetched {
+		err := e.fetchRowChunks()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.initPointers()
+		e.initCompareFuncs()
+		allCols := e.tryBuildKeyColumns()
+		if allCols {
+			sort.Slice(e.rowPointers, e.keyColumnsLess)
+		} else {
+			err = e.buildKeyChunks()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			sort.Slice(e.rowPointers, e.keyChunksLess)
+		}
+		e.fetched = true
+	}
+	for chk.NumRows() < chunk.MaxCapacity {
+		if e.Idx >= len(e.rowPointers) {
+			return nil
+		}
+		rowPtr := e.rowPointers[e.Idx]
+		chk.AppendRow(0, e.rowChunks[rowPtr.chkIdx].GetRow(int(rowPtr.rowIdx)))
+		e.Idx++
+	}
+	return nil
+}
+
+func (e *SortExec) fetchRowChunks() error {
+	fields := e.schema.GetTypes()
+	for {
+		chk := chunk.NewChunk(fields)
+		err := e.children[0].NextChunk(chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		rowCount := chk.NumRows()
+		if rowCount == 0 {
+			break
+		}
+		e.rowChunks = append(e.rowChunks, chk)
+		e.totalCount += rowCount
+	}
+	return nil
+}
+
+func (e *SortExec) initPointers() {
+	e.rowPointers = make([]rowPointer, 0, e.totalCount)
+	for chkIdx, rowChk := range e.rowChunks {
+		for rowIdx := 0; rowIdx < rowChk.NumRows(); rowIdx++ {
+			e.rowPointers = append(e.rowPointers, rowPointer{chkIdx: uint32(chkIdx), rowIdx: uint32(rowIdx)})
+		}
+	}
+}
+
+func (e *SortExec) initCompareFuncs() {
+	e.keyCmpFuncs = make([]chunk.CompareFunc, len(e.ByItems))
+	for i := range e.ByItems {
+		keyType := e.ByItems[i].Expr.GetType()
+		e.keyCmpFuncs[i] = chunk.GetCompareFunc(keyType)
+	}
+}
+
+func (e *SortExec) tryBuildKeyColumns() bool {
+	keyColumns := make([]int, 0, len(e.ByItems))
+	for _, by := range e.ByItems {
+		if col, ok := by.Expr.(*expression.Column); ok {
+			keyColumns = append(keyColumns, col.Index)
+		} else {
+			return false
+		}
+	}
+	e.keyColumns = keyColumns
+	return true
+}
+
+// keyColumnsLess is the less function for key columns.
+func (e *SortExec) keyColumnsLess(i, j int) bool {
+	ptrI := e.rowPointers[i]
+	ptrJ := e.rowPointers[j]
+	rowI := e.rowChunks[ptrI.chkIdx].GetRow(int(ptrI.rowIdx))
+	rowJ := e.rowChunks[ptrJ.chkIdx].GetRow(int(ptrJ.rowIdx))
+	for i, colIdx := range e.keyColumns {
+		cmpFunc := e.keyCmpFuncs[i]
+		cmp := cmpFunc(rowI, colIdx, rowJ, colIdx)
+		if e.ByItems[i].Desc {
```

Comment:
How about passing `bool desc` as one of the parameter of the comparers ?

Reply:
Sometimes we only need to compare equal, `Desc` is not needed.